### DISCUSSION TITLE
Colocated files could accept custom extensions on the new app folder

### DISCUSSION THREAD

**Discussion by jhortale** [vercel/next.js#44683] [Ideas]
### Describe the feature you'd like to request

It would be great if we have on the new app folder some configuration to accept pageExtensions, layoutExtension and all the other colocated files with its own name and extensions to avoid having so many files with the same file name. 

### Describe the solution you'd like

Basically the idea of extensions to have 
```
app
│   app.layout.tsx
│   app.page.tsx
└───Dashboard
    │   Dashboard.error.tsx
    │   Dashboard.head.tsx
    │   Dashboard.layout.tsx
    │   Dashboard.loading.tsx
    │   Dashboard.not-found.tsx
    │   Dashboard.page.tsx
    │   Dashboard.stories.tsx
    │   Dashboard.template.tsx
    │   Dashboard.test.tsx
```

### Describe alternatives you've considered

it would be something similar of what already exists for page extensions but applied to all the other colocated files into the app folder.

```js
{
experimental: {
   ...
    appDir: true,
    appDirExtensions: true,
  },
  pageExtensions: [
    'page.tsx',
  ],
  layoutExtensions: [
    'layout.tsx',
  ],
  errorExtensions: [
    'error.tsx',
  ],
  ...
}

**Comment by GabenGar** [comment]
Why not include `app` in the name also? It's not obvious from the filename this is a nextjs app dashboard, not just any dashboard. Isn't it how file system paths work?

**Comment by kkckkc** [comment]
I think this is a brilliant idea. It's one of the small, but very annoying issues, with using Next.js in a large codebase. The main issue this solves is usability in your editor/IDE, e.g. navigating to a file by name (Cmd-P in vscode) as well as title on editor tabs.

**Reply by GabenGar** [reply]
So in a large codebase you include every relevant detail of a file (domain, package, category, scope, main export, additional exports) into its name?

**Reply by kkckkc** [reply]
I guess large is a relative term, but I've found in practice that just a simple one level naming works well with up to at least ~100 pages (but you can add "domain" or "group" if needed), e.g.

```
Home.page.tsx
Category.page.tsx
Product.page.tsxt
AccountLogin.page.tsx
AccountAddressBook.page.tsx
```

Not sure what you mean by encoding main export and additional export. I don't think this is all or nothing, but having the ability to add *some* more context into the filenames is helpful IMHO

**Reply by GabenGar** [reply]
The context of filenames is their directory path. Using filenames as a hashtable is only tolerable to a point and gets worse once you need a shared abstract logic between files but non-shareable implementations of it, aka pages.
Also your naming example is wrong because `account login` is not a thing, since account by definition is already logged in and the only path to log in to account is from unknown state.

**Reply by kkckkc** [reply]
I guess this is a discussion about theory vs practice. I agree that we should not encode the full path in the filename. The only thing I'm asking for is the option to have more descriptive filenames - I'm not suggesting this is mandatory, but just to provide the option.  It will certainly help me and my team in practice The above setup works well with an internal large project with more than 100 developers around the world.

**Reply by GabenGar** [reply]
> I guess this is a discussion about theory vs practice.

Yeah vercel are some literal whos with no experience writing production-ready software.

> but just to provide the option

"An option" will complicate path resolution logic for everyone, introduce complexity with duplicate paths (like the old `pages` folder had with filenames and folders with index files) and a runtime performance tax due to extra lookups.

>The above setup works well with an internal large project with more than 100 developers around the world.

India isn't the entire world though and the amount of developers isn't indicative of project size.

**Reply by kkckkc** [reply]
Hi again. It seems my communication skills are not great, I did not intent to cause such hard feelings - my apologies. 

I have nothing but respect for the Vercel team and did not in any way mean to imply a lack of experience with large projects. The point about "theory vs practice" was related to your point about the path already containing the full context (which is obviously correct) vs an IDE/editor often just showing the filename in the editor Tabs (so the context is "hard" to use/see in practice) - I should have been more clear about this - and it's not a very good analogy. Also I'm not sure about the reference to India, my team is mostly Europe and US and I'm based in Sweden.

Anyhow, for me and my team this is a valuable feature - but I can solve it myself (I have a patch) and it's possible to just use symlinks instead. So if no one else believes this is a useful feature, I'm happy to continue with my workaround.

Thanks for the point about duplicate paths, that is a valid argument that I didn't think about - and I see that this may cause more harm than good in the general case.
